gcp_variant_transforms/beam_io/vcf_parser.py

```diff
@@ -504,6 +570,224 @@ def _get_variant_calls(self, record, formats):
+      self._to_child.flush()
+      return self._convert_to_variant(next(self._vcf_reader))
+    except (MemoryError, IOError, ValueError) as e:
+      logging.warning('VCF record read failed in %s for line %s: %s',
+                      self._file_name, data_line, str(e))
+      return MalformedVcfRecord(self._file_name, data_line, str(e))
+
+  def _verify_record(self, record):
+    if record.start < 0:
+      raise ValueError('Start position is incorrect.')
+    if record.stop < 0:
+      raise ValueError('End position is incorrect.')
+
+  def _convert_to_variant(self, record):
+    # type: (libcbcf.VariantRecord) -> Variant
+    """Converts the PySAM record to a :class:`Variant` object.
+
+    Args:
+      record: An object containing info about a variant.
+
+    Returns:
+      A :class:`Variant` object from the given record.
+
+    Raises:
+      ValueError: if `record` is semantically invalid.
+    """
+    self._verify_record(record)
+    return Variant(
+        reference_name=record.chrom.encode('utf-8'),
+        start=record.start,
+        end=record.stop,
+        reference_bases=self._convert_field(record.ref),
+        alternate_bases=list(record.alts) if record.alts else [],
+        names=record.id.split(';') if record.id else [],
+        quality=self._parse_to_numeric(record.qual) if record.qual else None,
+        filters=(None if not list(record.filter.keys()) else
+                 list(record.filter.keys())),
+        info=self._get_variant_info(record),
+        calls=self._get_variant_calls(record.samples))
+
+  def _get_variant_info(self, record):
+    # type: (libcbcf.VariantRecord) -> Dict[str, Any]
+    info = {}
+    for k, v in list(record.info.items()):
+      if k != END_INFO_KEY:
+        if isinstance(v, tuple):
+          info[k] = list(map(self._convert_field, v))
+        else:
+          # If a field was not provided in the header, make it a list to
+          # follow the PyVCF standards.
+          info[k] = (
+              self._convert_field(v) if k in self._original_info_list else
+              [self._convert_field(v)])
+
+    return info
+
+  def _parse_to_numeric(self, value):
+    # type: (Any) -> Any
+    if isinstance(value, str) and value.isdigit():
+      return int(value)
+    else:
+      try:
+        return self._parse_float(float(value))
+      except ValueError:
+        # non float
+        if isinstance(value, unicode):
+          return value.encode('utf-8')
+        return value
+
+  def _parse_float(self, value):
+    # type: (Any) -> Any
+    precise_value = float("{:0g}".format(value))
+    if precise_value.is_integer():
+      return int(precise_value)
+    return precise_value
+
+  def _convert_field(self, value, numeric=True):
+    # type: (Any) -> Any
+    # PySam currently doesn't recognize '.' value for String fields as missing.
+    if value == '.' or value == b'.' or not value:
+      return None
+    elif isinstance(value, unicode):
```

Comment:
Again I'm not sure what this is doing - what is a 'unicode' object in python and why does it need special handling?

Reply:
Well unicode an 'unencoded' sring. PySam sometimes returns unicode values, sometimes str-s in utf-8 encoding, so I encode all to utf-8, to stay consistent.

Reply:
@tneymanov But do you know why that is?  There is a very specific reason.

Reply:
Did some analysis over the valid-4.1-large.vcf file - seems like whenever PySam adds a string of its own (like '.' or 'PASS'), or when an INFO field is set to be string, PySam return unicode. In fact the only time it returns str type is for REF alleles...

Is there any other reason? Or any insight why it works like that above?

Reply:
Pretty good first analysis!  To make a short story short, it has to have some consistency among Python versions.  Strings in Python 3 are unicode, while in Python 2 you have to make them unicode.  See: https://portingguide.readthedocs.io/en/latest/strings.html

That's why they mentioned it specifically starting with release 0.8.1 for Python 3 in https://pysam.readthedocs.io/en/latest/release.html#release-0-8-1

That's why you see things like this in [libcutils.pyx](https://github.com/pysam-developers/pysam/blob/master/pysam/libcutils.pyx#L157-L168):

```Python
cdef force_str(object s, encoding=TEXT_ENCODING):
    """Return s converted to str type of current Python
    (bytes in Py2, unicode in Py3)"""
    if s is None:
        return None
    if PY_MAJOR_VERSION < 3:
        return s
    elif PyBytes_Check(s):
        return s.decode(encoding)
    else:
        # assume unicode
        return s
```

Or in [libcbcf.pyx](https://github.com/pysam-developers/pysam/blob/master/pysam/libcbcf.pyx#L151-L154):

```Python
    if PY_MAJOR_VERSION < 3:
        val = s
    else:
        val = PyUnicode_DecodeUTF8(s, strlen(s), NULL)
```